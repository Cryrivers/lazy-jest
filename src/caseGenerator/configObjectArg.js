// @flow
import type { Case, ArgConfig } from '../index.flow';
import { default as enumerateCases, appendArgCases } from './enumerateCases';

type ObjectCase = { [key: string]: Case };

/**
 * Append a property to each object of a list. Similar to "setEach" in some functional programming lib.
 * @private
 * @param {string} key property name
 * @return {extendObjectCase~callback}
 * @example
 * const cases = extendObjectCase('a')([{}], 1);  // Result: [{ a: 1 }]
 * extendObjectCase('b')(cases, '');  // Result: [{ a: 1, b: '' }]
 * extendObjectCase('c')([{ a: 1 }, { b: '' }], []);  // Result: [{ a: 1, c: [] }, { b: '', c: [] }]
 */
const extendObjectCase = (key: string) => {
  return (exsistingCases: ObjectCase[], nextArgCase: Case): ObjectCase[] => {
    if (exsistingCases.length) {
      return exsistingCases.map(c => {
        return { ...c, [key]: nextArgCase };
      });
    }
    return [{ [key]: nextArgCase }];
  };
};
/**
 * @callback extendObjectCase~callback
 * @param {Array<Object>} cases
 * @param {*} nextArgCase
 * @return {Array<Object>}
 */

const getTestCases = (key: string) => enumerateCases(extendObjectCase(key));
const appendTestCases = (key: string) => appendArgCases(extendObjectCase(key));

/**
 * A helper to generate object test cases by provided configurations
 * @param {ArgConfig} propsConfig
 * @return {{validCases: Array<Object>, invalidCases: Array<Object>}}
 * @example
 * makeObjectArg([
 *  {
 *    name: 'a',
 *    validCases: [1, 2],
 *    invalidCases: [NaN, 0]
 *  },
 *  {
 *    name: 'b',
 *    validCases: ['a'],
 *    invalidCases: ['', 1],
 *    optional: true
 *  }
 * ]);
 * // Result:
 * // {
 * //   validCases: [
 * //     { a: 1 },
 * //     { a: 2 },
 * //     { a: 1, b: a }
 * //   ],
 * //   invalidCases: [
 * //     { a: NaN },
 * //     { a: 0 },
 * //     { a: 1, b: '' },
 * //     { a: 1, b: 1 }
 * //   ]
 * // }
 */
const makeObjectArg = (propsConfig: ArgConfig[]) => {
  const compulsoryProps = propsConfig.filter(conf => !conf.optional);
  const optionalProps = propsConfig.filter(conf => conf.optional);
  let compulsoryValidCases = [];
  let compulsoryInvalidCases = [];
  compulsoryProps.forEach(({ name }, i) => {
    const getCases = getTestCases(name);
    compulsoryInvalidCases = compulsoryInvalidCases.concat(
      getCases(compulsoryProps, i)
    );
    compulsoryValidCases = compulsoryValidCases.concat(
      getCases(compulsoryProps)
    );
  });
  let optionalValidCases = [];
  let optionalInvalidCases = [];
  optionalProps.forEach(({ name, invalidCases, validCases }) => {
    const append = appendTestCases(name);
    optionalInvalidCases = optionalInvalidCases.concat(
      append(compulsoryValidCases, invalidCases)
    );
    optionalValidCases = optionalValidCases.concat(
      append(compulsoryValidCases, validCases)
    );
  });
  return {
    validCases: [...compulsoryProps, ...optionalValidCases],
    invalidCases: [...compulsoryInvalidCases, ...optionalInvalidCases]
  };
};

export default makeObjectArg;
