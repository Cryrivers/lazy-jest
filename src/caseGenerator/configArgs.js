// @flow
import type { Case, ArgConfig } from '../index.flow';
import configObjectArg from './configObjectArg';

/**
 * Validate argConfig and adjust some values
 * @private
 * @param {ArgConfig} argConfig
 * @return {ArgConfig}
 */
const validateArgConfig = (argConfig: ArgConfig) => {
  let { validCases, invalidCases, optional, name } = argConfig;
  if (optional) {
    invalidCases = invalidCases
      ? invalidCases.filter(c => typeof c !== 'undefined')
      : void 0;
    if (!validCases.includes(void 0)) {
      validCases = [...validCases, void 0];
    }
  }
  return {
    optional: !!optional,
    name: name || 'argument',
    validCases,
    invalidCases
  };
};

/**
 * Helper for generate ArgConfig
 * @protected
 */
export class ArgsConfig {
  value: ArgConfig[];

  /**
   * @param {ArgConfig[]} initialConfig
   */
  constructor(initialConfig?: ArgConfig[]) {
    this.value = Array.isArray(initialConfig) ? initialConfig.map(validateArgConfig) : [];
  }
  /**
   * Add an argument
   * @param {string} name 
   * @param {Array.<*>} validCases valid test cases for this argument
   * @param {Object.<string, *>} [opts] other options
   * @prop {boolean} [opts.optional] flag to indicate whether this argument is optional
   * @prop {Array.<*>} [opts.invalidCases] invalid test cases for thsi argument
   */
  arg(
    name: string,
    validCases: Case[],
    opts?: { optional?: boolean, invalidCases?: Case[] }
  ) {
    this.value.push(
      validateArgConfig({
        ...(opts || {}),
        name,
        validCases
      })
    );
    return this;
  }
  /**
   * Add an object type argument
   * @param {string} name 
   * @param {ArgsConfig} propsConfig 
   * @param {Object.<string, *>} [opts]
   * @prop {boolean} [opts.optional] 
   */
  objectArg(name: string, propsConfig: ArgsConfig, opts?: { optional?: boolean }) {
    this.value.push({
      ...configObjectArg(propsConfig),
      ...(opts || {}),
      name
    });
    return this;
  }
}

/**
 * @memberof ArgsConfig
 * @instance
 * @example
 * // example ArgConfig for `func(a, b) {}`
 * configArgs()
 * .arg('a', [0, 1, 2])
 * .arg('b', [-1, -2], { optional: true, invalidCases: [0] });
 */
const configArgs = (initialConfig?: ArgConfig[]) => new ArgsConfig(initialConfig);
export default configArgs;
