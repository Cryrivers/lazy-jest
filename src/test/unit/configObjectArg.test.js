import configObjectArg from '../../caseGenerator/configObjectArg';
import configArgs from '../../caseGenerator/configArgs';

describe('configObjectArg', () => {
  test('empty object', () => {
    expect(configObjectArg(configArgs())).toMatchSnapshot();
  });

  test('single property', () => {
    expect(configObjectArg(configArgs().arg('a', [1]))).toMatchSnapshot();
  });

  test('single property has invalid cases', () => {
    expect(configObjectArg(configArgs().arg('a', [1], { invalidCases: [null] }))).toMatchSnapshot();
  });

  test('single optional property', () => {
    expect(configObjectArg(configArgs().arg('a', [1], { optional: true }))).toMatchSnapshot();
  });

  test('multiple properties, no optional', () => {
    expect(configObjectArg(
      configArgs()
      .arg('a', [1])
      .arg('b', [0])
    )).toMatchSnapshot();
  });

  test('multiple properties, has optional', () => {
    expect(configObjectArg(
      configArgs()
      .arg('a', [1])
      .arg('b', [0], { optional: true })
    )).toMatchSnapshot();
  });
  
  test('multiple properties, has invalid cases', () => {
    expect(configObjectArg(
      configArgs()
      .arg('a', [1])
      .arg('b', [0], { invalidCases: [null] })
    )).toMatchSnapshot();
  });
});