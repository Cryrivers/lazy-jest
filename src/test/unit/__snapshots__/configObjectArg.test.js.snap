// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`configObjectArg empty object 1`] = `
Object {
  "invalidCases": Array [],
  "validCases": Array [],
}
`;

exports[`configObjectArg multiple properties in array 1`] = `
Object {
  "invalidCases": Array [],
  "validCases": Array [
    Object {
      "name": "a",
      "validCases": Array [
        1,
      ],
    },
    Object {
      "name": "b",
      "validCases": Array [
        0,
      ],
    },
  ],
}
`;

exports[`configObjectArg multiple properties, has invalid cases 1`] = `
Object {
  "invalidCases": Array [
    Object {
      "b": null,
    },
  ],
  "validCases": Array [
    Object {
      "invalidCases": undefined,
      "name": "a",
      "optional": false,
      "validCases": Array [
        1,
      ],
    },
    Object {
      "invalidCases": Array [
        null,
      ],
      "name": "b",
      "optional": false,
      "validCases": Array [
        0,
      ],
    },
  ],
}
`;

exports[`configObjectArg multiple properties, has optional 1`] = `
Object {
  "invalidCases": Array [],
  "validCases": Array [
    Object {
      "invalidCases": undefined,
      "name": "a",
      "optional": false,
      "validCases": Array [
        1,
      ],
    },
    Object {
      "a": 1,
      "b": 0,
    },
    Object {
      "a": 1,
      "b": undefined,
    },
  ],
}
`;

exports[`configObjectArg multiple properties, no optional 1`] = `
Object {
  "invalidCases": Array [],
  "validCases": Array [
    Object {
      "invalidCases": undefined,
      "name": "a",
      "optional": false,
      "validCases": Array [
        1,
      ],
    },
    Object {
      "invalidCases": undefined,
      "name": "b",
      "optional": false,
      "validCases": Array [
        0,
      ],
    },
  ],
}
`;

exports[`configObjectArg optional property has invalid cases 1`] = `
Object {
  "invalidCases": Array [
    Object {
      "a": null,
    },
  ],
  "validCases": Array [
    Object {
      "a": 1,
    },
    Object {
      "a": undefined,
    },
  ],
}
`;

exports[`configObjectArg single optional property 1`] = `
Object {
  "invalidCases": Array [],
  "validCases": Array [
    Object {
      "a": 1,
    },
    Object {
      "a": undefined,
    },
  ],
}
`;

exports[`configObjectArg single property 1`] = `
Object {
  "invalidCases": Array [],
  "validCases": Array [
    Object {
      "invalidCases": undefined,
      "name": "a",
      "optional": false,
      "validCases": Array [
        1,
      ],
    },
  ],
}
`;

exports[`configObjectArg single property has invalid cases 1`] = `
Object {
  "invalidCases": Array [
    Object {
      "a": null,
    },
  ],
  "validCases": Array [
    Object {
      "invalidCases": Array [
        null,
      ],
      "name": "a",
      "optional": false,
      "validCases": Array [
        1,
      ],
    },
  ],
}
`;
